Add missing RBI definitions for core

Morriar opened the pull request:
### Motivation

More RBI definitions for `core/`. No signatures.

### Test plan
<!-- If you did not write tests for this change, replace the message below explaining why not. Why we should be confident this change is correct? If you changed the website, please include a screenshot of the proposed changes. -->

See included automated tests.

jez commented:
https://go/builds/bui_HMLf3YJpzjxOoh

jez commented:
The problem I'm seeing when running this on Stripe's codebase:

The `array.rbi` changes shadow the superclass methods from Enumerable, making something that was previously typed now untyped on this branch.

For example `Array#max` and `Array#min` are no longer typed.

We have a bunch of code at stripe that is doing `T.must(xs.max)`, but `T.must` requires that the wrapped expression be nilable, not untyped.

Do you mind either backing out the array.rbi changes, or auditing them to not shadow the corresponding methods in enumerable.rbi?

Morriar commented:
@jez I rebased this on master, solved the conflicts in exp files and removed the overloads from Enumerable. Tests are green, let me know if another method is conflicting.

jez commented:
https://go/builds/bui_HMad69njteAtPq

jez commented on rbi/core/array.rbi:
````diff
@@ -483,6 +483,43 @@ class Array < Object
+  # [ "a", "c", 7 ] == [ "a", "c", 7 ]     #=> true
+  # [ "a", "c", 7 ] == [ "a", "d", "f" ]   #=> false
+  # ```
+  def ==(_); end
````
Every `def ==(` here that doesn't have a type overrides the `T::Boolean` return type of `BasicObject#==`

jez replied:
(there are a handful of `==` methods in this PR that need to be typed or deleted)

jez commented on rbi/core/hash.rbi:
```diff
@@ -1032,6 +1266,58 @@ class Hash < Object
+  # [`Hash#update`](https://docs.ruby-lang.org/en/2.6.0/Hash.html#method-i-update)
+  # is an alias for
+  # [`Hash#merge!`](https://docs.ruby-lang.org/en/2.6.0/Hash.html#method-i-merge-21).
+  def update(_); end
```
Our reflection-generated RBI says that this is variadic, not unary:

```
❯ irb
irb(main):001:0> Hash.instance_method(:update).parameters
=> [[:rest]]
```

jez commented on rbi/core/hash.rbi:
```diff
@@ -570,6 +713,29 @@ class Hash < Object
+  #
+  # See also
+  # [`Enumerable#include?`](https://docs.ruby-lang.org/en/2.6.0/Enumerable.html#method-i-include-3F)
+  def include?(_); end
```
This overrides and untypes `Enumerable#include?`

jez commented on rbi/core/array.rbi:
````diff
@@ -1177,6 +1298,27 @@ class Array < Object
+  # a.index("z")              #=> nil
+  # a.index {|x| x == "b"}    #=> 1
+  # ```
+  def find_index(*_); end
````
This overrides and untypes `Enumerable#find_index`

Morriar commented:
@jez I removed the remaining conflicts and fixed the signature for `update`. Let me know if something else is blocking.

jez commented:
https://go/builds/bui_HMbRTRqpm4RNA3

jez commented on rbi/core/hash.rbi:
````diff
@@ -147,6 +147,73 @@ class Hash < Object
+  # h2 < h1    #=> false
+  # h1 < h1    #=> false
+  # ```
+  def <(_); end
````
```suggestion
  sig {params(_: T.untyped).returns(T::Boolean)}
  def <(_); end
```

jez commented on rbi/core/hash.rbi:
````diff
@@ -147,6 +147,73 @@ class Hash < Object
+  # h2 = { "c" => 2, "a" => 1 }
+  # h1 == h2   #=> true
+  # ```
+  def ==(_); end
````
```suggestion
  sig {params(_: T.untyped).returns(T::Boolean)}
  def ==(_); end
```

jez commented on rbi/core/hash.rbi:
````diff
@@ -147,6 +147,73 @@ class Hash < Object
+  # h2 > h1    #=> true
+  # h1 > h1    #=> false
+  # ```
+  def >(_); end
````
```suggestion
  sig {params(_: T.untyped).returns(T::Boolean)}
  def >(_); end
```

jez commented on rbi/core/hash.rbi:
````diff
@@ -147,6 +147,73 @@ class Hash < Object
+  # h2 >= h1   #=> true
+  # h1 >= h1   #=> true
+  # ```
+  def >=(_); end
````
```suggestion
  sig {params(_: T.untyped).returns(T::Boolean)}
  def >=(_); end
```
